Ability to punch objects

emmanuelbassil opened the pull request:
🆑 emmanuelbassil
bugfix: Fixes being unable to punch spiderlings to death
tweak: Can now damage any object with health using your fists of justice; similar to existing behavior with melee weapons.
/🆑 

Wanted to fix the bug where you can't punch spiderlings. Noticed that there are plenty of things you can't punch that you can nonetheless damage with melee weapons. Fixed this most grievous oversight.

SierraKomodo commented on code/game/objects/objs.dm:
```diff
@@ -243,8 +243,18 @@
+		var/datum/unarmed_attack/attack = assailant.get_unarmed_attack(src)
+		var/damage = attack.damage + rand(1,5)
+		playsound(loc, attack.attack_sound, 25, 1, -1)
+		assailant.visible_message(SPAN_WARNING("[assailant] [pick(attack.attack_verb)] [src]!"))
```
```suggestion
        var/attack_verb = [pick(attack.attack_verb)]
		assailant.visible_message(
            SPAN_WARNING("\The [assailant] [attack_verb] \the [src]!"),
            SPAN_WARNING("You [attack_verb] \the [src]!)
        )
```

SierraKomodo commented on code/game/objects/objs.dm:
```diff
@@ -243,8 +243,18 @@
+
+	if (ishuman(user) && user.a_intent == I_HURT && get_max_health())
+		var/mob/living/carbon/human/assailant = user
+		var/datum/unarmed_attack/attack = assailant.get_unarmed_attack(src)
```
This needs to be null-checked - There's the _potential_ it could be null, looking at the proc itself.

SierraKomodo commented on code/game/objects/objs.dm:
```diff
@@ -243,8 +243,18 @@
+		var/mob/living/carbon/human/assailant = user
+		var/datum/unarmed_attack/attack = assailant.get_unarmed_attack(src)
+		var/damage = attack.damage + rand(1,5)
+		playsound(loc, attack.attack_sound, 25, 1, -1)
```
```suggestion
		playsound(loc, attack.attack_sound, 25, TRUE, -1)
```

SierraKomodo commented on code/game/objects/objs.dm:
```diff
@@ -243,8 +243,18 @@
+		assailant.visible_message(SPAN_WARNING("[assailant] [pick(attack.attack_verb)] [src]!"))
+		assailant.do_attack_animation(src)
+		assailant.setClickCooldown(DEFAULT_ATTACK_COOLDOWN)
+		damage_health(damage, attack.get_damage_type(), attack.damage_flags())
```
Prevent double-actions from a single click.

```suggestion
		damage_health(damage, attack.get_damage_type(), attack.damage_flags())
        return
```

emmanuelbassil commented:
Fixed; ready again.

emmanuelbassil commented:
Please hold on this; will add an isequipped check first. Not today; did enough coding.

emmanuelbassil commented:
Add isitem check; ready for review again

SierraKomodo commented on code/game/objects/objs.dm:
```diff
@@ -243,8 +243,25 @@
+	if (ishuman(user) && !isitem(src) && user.a_intent == I_HURT && get_max_health())
+		var/mob/living/carbon/human/assailant = user
+		var/datum/unarmed_attack/attack = assailant.get_unarmed_attack(src)
+		if (!attack) return
```
```suggestion
		if (!attack)
            return ..()
```

SierraKomodo commented on code/game/objects/objs.dm:
```diff
@@ -243,8 +243,25 @@
+		var/damage = attack.damage + rand(1,5)
+		var/attack_verb = "[pick(attack.attack_verb)]"
+		playsound(loc, attack.attack_sound, 25, TRUE, -1)
+		assailant.visible_message(
+				SPAN_WARNING("\The [assailant] [attack_verb] \the [src]!"),
+				SPAN_WARNING("You [attack_verb] \the [src]!")
+				)
+		assailant.do_attack_animation(src)
+		assailant.setClickCooldown(DEFAULT_ATTACK_COOLDOWN)
+		damage_health(damage, attack.get_damage_type(), attack.damage_flags())
+		return
```
This should check `can_damage_health()` and provide a separate message that indicates the attack failed. See the implementation for `/atom/proc/use_weapon()` for an example.

emmanuelbassil commented:
Done. 
Also made it cause damage to the hand if !can_damage_health fails; similar to how punching walls behaves.

SierraKomodo commented on code/game/objects/objs.dm:
```diff
@@ -243,8 +243,36 @@
+
+		if (!can_damage_health(damage, attack.get_damage_type()))
+			playsound(loc, attack.attack_sound, 25, TRUE, -1)
+			user.visible_message(
+				SPAN_WARNING("\The [user] hits \the [src], but doesn't even leave a dent!"),
+				SPAN_WARNING("You hit \the [src], but cause no visible damage.")
+			)
+			user.apply_damage(3, DAMAGE_BRUTE, user.hand ? BP_L_HAND : BP_R_HAND)
+			return TRUE
```
If the user's taking damage for it, the message to the user should reflect that.


```suggestion
			user.visible_message(
				SPAN_WARNING("\The [user] hits \the [src], but doesn't even leave a dent!"),
				SPAN_WARNING("You hit \the [src], but cause no visible damage and hurt yourself!")
			)
			user.apply_damage(3, DAMAGE_BRUTE, user.hand ? BP_L_HAND : BP_R_HAND)
			return TRUE
```

emmanuelbassil committed:
Ability to punch objects

Fixes

Auto stash before rebase of "refs/heads/AssortedImp"

Fixes

Auto stash before rebase of "refs/heads/Squish"

emmanuelbassil commented:
Handled

SierraKomodo reviewed (approved):
Doesn't work with objects that already have an attack_hand override, i.e. closets, but that'll require a deeper refactor to resolve.

Otherwise, works as-is.
